Modules: Concurrency module context fix

pm-nikhil-vaidya opened the pull request:
https://github.com/prebid/prebid-server/issues/4239

pm-nikhil-vaidya committed:
Modules: Concurrency issue with module context fix

bsardo commented on hooks/hookstage/invocation.go:
```diff
@@ -28,11 +29,75 @@ type ModuleInvocationContext struct {
 // We use interface as we do not know exactly how the modules will use their inner context.
-type ModuleContext map[string]interface{}
+type ModuleContext struct {
+	sync.RWMutex
```
Discussed in eng subcommittee. Would it be better to use a sync map?

pm-nikhil-vaidya replied:
@scr-oath
Agree! Using mutex basically introduce two issues:

1. It basically take a performance hit and make things slower
2. Particularly in two hook stage bidderRequest and rawBidderResponse, a lock hold by a slow bidder may slow down entire bidder execution.

The main reason of panic, when a module in this stages try to update module context passed in module invocation context. Ideally module should only read from the moduleInvocation context and then updated module context should be set in result instead of directly updating the instance.

Here, we are just moving locking mechanism from above mentioned hook stage to stage execution so we are only solving second issue here.

Instead we can create a copy of module Context, and send copy while executing the stage. This might solve the problem and we can avoid concurrent read and write to same map and whatever data needs to be updated in module context should be updated via result.

Thoughts?

bsardo commented:
@postindustria-code can you please review?

scr-oath commented:
@pm-nikhil-vaidya - Happy New Year - can you take a look at the diff and see if you are comfortable applying it?
https://github.com/prebid/prebid-server/pull/4603#discussion_r2627751038

pm-nikhil-vaidya committed:
Switched to more performant solution while adding module contexts

pm-nikhil-vaidya commented:
@scr-oath I have implemented the solution you suggested as it removes intermediate copy while copying the module context and removes extra lock mechanism while reading. I have just added a check in the Insert() function in case of map map is empty.

```go
func (mc *ModuleContext) Insert(seq iter.Seq2[string, any]) {
	if mc == nil {
		return
	}

	mc.Lock()
	defer mc.Unlock()

	if mc.data == nil {
		mc.data = make(map[string]any)
	}

	maps.Insert(mc.data, seq)
}
```
You can review and let me know if any more changes required

pm-nikhil-vaidya committed:
Reverted to GetAll and SetAll methods to update the ModuleCtxs

scr-oath committed:
Merge branch 'master' into module-context-concurrency

bsardo committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya committed:
Removed unused methods for module context update

pm-nikhil-vaidya commented on hooks/hookstage/invocation.go:
```diff
@@ -28,11 +31,102 @@ type ModuleInvocationContext struct {
+var emptyMapIter iter.Seq2[string, any] = func(yield func(string, any) bool) {}
+
+// All returns an iterator over key-value pairs from the module context with read lock held
+func (mc *ModuleContext) All() iter.Seq2[string, any] {
```
Done

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

scr-oath reviewed (dismissed):
All three previous change requests have been addressed:

1. ✅ `make(map[string]any, len(data))` — properly sized allocation in `SetAll` when `data` is nil
2. ✅ `All()` iterator removed — eliminates deadlock risk from holding lock across iteration
3. ✅ `Insert()` removed — same concern addressed

**Concurrency analysis of `put()`:** The copy-based approach (`existingCtx.SetAll(mCtx.GetAll())`) is safe. Go evaluates `mCtx.GetAll()` first (acquires/releases `mCtx.RLock`), then passes the returned plain map to `existingCtx.SetAll()` (acquires/releases `existingCtx.Lock`). No two `ModuleContext` locks are ever held simultaneously. No reverse-order deadlock is possible since `ModuleContext` methods never reference back to `moduleContexts`.

**nitpick (non-blocking):** `SetAll` comment says "replaces all data" but `maps.Copy` merges into the existing map (existing keys not in `data` are preserved). The merge behavior is correct for this use case (accumulating context across stages), but the comment could say "merges data into the context" to be precise.

LGTM — the RWMutex + copy-based approach is sound, modules are all updated correctly, and tests are comprehensive.

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

scr-oath reviewed (changes requested):
Re-reviewing after my previous approval was dismissed. The only delta since my last review is the v3→v4 import path bump from the master merge — the logic is unchanged.

**One blocking issue found**: data race in `Get`/`GetAll` where `mc.data` is read outside the lock.

Previous change-requests (sized allocation in `SetAll`, removed `All()` iterator, removed `Insert()`) are all still correctly addressed. The concurrency design is sound aside from the pre-lock nil check.

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya commented on hooks/hookstage/invocation.go:
```diff
@@ -28,11 +30,71 @@ type ModuleInvocationContext struct {
+	return result
+}
+
+// SetAll replaces all data in the context
```
Done

pm-nikhil-vaidya commented on hooks/hookexecution/context.go:
```diff
@@ -43,24 +43,27 @@ func (ctx executionContext) getModuleContext(moduleName string) hookstage.Module
-	mc.ctxs[moduleName] = newCtx
+
+	// Add new data to existing context
+	existingCtx.SetAll(mCtx.GetAll())
```
This will partially address the nested locking issue. Since setAll also acquires a lock, nested locking won’t be completely eliminated. Therefore, we can retain the existing implementation.

pm-nikhil-vaidya committed:
Moved data nil check inside lock protection

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency

pm-nikhil-vaidya committed:
Used synced moduleCtx for device detection module

scr-oath commented on modules/scientiamobile/wurfl_devicedetection/module.go:
```diff
@@ -101,7 +101,11 @@ func (m Module) HandleRawAuctionHook(
-	rawHeaders, ok := invocationCtx.ModuleContext[wurflHeaderCtxKey].(map[string]string)
+	headers, ok := invocationCtx.ModuleContext.Get(wurflHeaderCtxKey)
+	if !ok {
+		return result, hookexecution.NewFailure("invalid module context type")
```
nitpick (non-blocking): error message is misleading on this branch.

`Get` returning `ok=false` means the key is missing, not that the value's type is wrong. The next branch (line 110) handles the actual type-assertion failure with the same string, so the two failure modes are no longer distinguishable.

Consider:
```go
headers, ok := invocationCtx.ModuleContext.Get(wurflHeaderCtxKey)
if !ok {
    return result, hookexecution.NewFailure("missing module context entry")
}
rawHeaders, ok := headers.(map[string]string)
if !ok {
    return result, hookexecution.NewFailure("invalid module context type")
}
```

scr-oath reviewed (approved):
Re-reviewing after the fix commits.

**The prior blocking race is fixed.** `Get` and `GetAll` in `hooks/hookstage/invocation.go` now check `mc.data == nil` *inside* the read lock, ordered behind `Set`/`SetAll`'s lazy init under the write lock. The `SetAll` doc comment was also corrected from "replaces" to "merges". The two CI-breaking `ModuleContext[asyncRequestKey]` map-indexes in `modules/scope3/rtd/module_test.go` are now `.Get(...)`. The new wurfl device-detection migration is a clean, mechanical port that mirrors the pattern used by the other modules in this PR.

**Verification:**
1. Race fix — `invocation.go` lines 58–65 (Get) and 85–92 (GetAll): `RLock` taken, then nil check, then map read. Symmetric on the write side at lines 72–78 (Set) and 100–106 (SetAll). No window remains.
2. `put()` in `hooks/hookexecution/context.go:63` — copy-based merge is still deadlock-free: `mCtx` and `existingCtx` are distinct `*ModuleContext` instances, and `ModuleContext` methods never call back into `moduleContexts`. Lock graph is acyclic.
3. CI fixes — both leftover map-index sites in `module_test.go` migrated; the assertion semantics (`assert.NotNil` on interface nil) are preserved because `Get` returns untyped-nil on miss.

**LGTM.** One non-blocking nit on the wurfl error message left as a line comment.

Reviewer notes:
- The two carryover non-blocking suggestions from my prior review (extract snapshot before `put`'s outer lock; helper for the duplicated scope3 `Get`+type-assert pattern) remain unaddressed. Author's call — neither blocks merge.

pm-nikhil-vaidya committed:
Merge branch 'master' into module-context-concurrency
